### DISCUSSION TITLE
Glossary / Lexicon

### DISCUSSION THREAD

**Discussion by taylor** [lfn-cnti/bestpractices#136] [General]
Please start adding terms here and defining them. We can create a PR to add to the repo glossary as needed and when there is agreement.

Please try to map to K8s terms as we are focused on K8s-based environments and need to collaborate with the community.

**Comment by jeffsaelens** [comment]
Lets attack the elephant in the room... CNF definition 

A cloud native network function (CNF) is a cloud native application that implements network functionality. A CNF consists of one or more microservices and has been developed using Cloud Native Principles including immutable infrastructure, declarative APIs, and a “repeatable deployment process”.

From the [TUG white paper](https://github.com/cncf/telecom-user-group/blob/master/whitepaper/cloud_native_thinking_for_telecommunications.md#1.4)

I'm sure there will be debate. I'll start by openly stating I'm against any definition that specifically binds us to an ETSI NFV paradigm, or simply states their Cloud Native VNFs.

**Reply by CsatariGergely** [reply]
I'm okay with this definition. Maybe we could add horizontal scalability, but not really important as the list is open ended anyways.

**Reply by electrocucaracha** [reply]
So another definition could be _An application which provides network functionality implementing/using Cloud Native principles._

**Reply by jeffsaelens** [reply]
So seems the definition is the same, just a rewording to make things more clear?

**Reply by electrocucaracha** [reply]
Something to clarify, a _Cloud Native application_ must implement all the Cloud Native principles or at least one of them?

**Reply by taylor** [reply]
Existing discussion for CNF definition https://github.com/cncf/cnf-wg/discussions/27

**Reply by taylor** [reply]
@fkautz said in zoom chat: 
> https://x.CNF.dev was formational to defining the initial CNF definition which led to the creation of the CNCF TUG

**Comment by gupta-alok** [comment]
Following are the terms and their definitions related to external network orchestration https://github.com/cncf/cnf-wg/discussions/118

- **Network Attachment**
A network attachment represents a network interface of a Kubernetes pod. 

  - **Primary Network Attachment**
Each pod has a mandatory primary network attachment. This kernel network interface (eth0) is injected by the default CNI.
  - **Secondary Network Attachment**
On top of the primary network attachment a K8s pod may have zero or more secondary network attachments. These are requested in the pod specification by means of network annotations referring to Network Attachment Definitions (NADs).
A meta CNI plugin, such as Multus, iterates over the pod’s network annotations and invokes the CNI plugin specified in each NAD to inject the requested secondary network attachment into the pod.
Secondary network attachments can take many forms, among others kernel network interfaces, PCI devices configured for use with DPDK, AF_XDP sockets, or user-space interfaces like memif.

- **Primary Network**
The network to which the K8s pods attach with their primary network interface (eth0). It includes the internal network interconnecting the K8s pods and any external networks with connectivity to the primary pod interfaces, either directly through IP routing, involving NAT and/or load-balancing. Today, K8s services reside on and are limited to the primary network.

- **Secondary Network**
A network, isolated from the primary network, that K8s pods can attach to through secondary network attachments. Typically, secondary networks have external connectivity through the DC Edge.

- **Overlay Network**
Overlay Network is a networking abstraction that creates multiple layers of networking on top of physical network. Through overlay networks, pods can communicate with each other within their cluster and with the gateway(s) connected to networks outside of the data center by creating virtual networks.

- **External Network**
A network connecting the K8s cluster nodes and pods with external entities, typically the DC Edge for connectivity with the WAN. 

- **Tenant Network**
The networks that carry tenant traffic (e.g. traffic to/from CNFs) are simply called Tenant Traffic Network. Tenant networks comprise primary and secondary networks. Tenant networks are typically isolated from infrastructure networks, such as the control network carrying the K8s cluster control plane or storage traffic.

- **Control Network** 
A Control Network is a L2 or L3 network used to exchange control information among different infrastructure nodes. 

- **DC-Gw/Site Router/DC Edge Router (PE + GW)**
The Data Center Gateway is a router through which the traffic entering or leaving to the data center, is routed. It faces on one side the data center network and on the other side a transport network. It may act as the DC Edge and may also be combined with DC Management Edge, WAN Edge and/or access edge. 

- **POD (Performance Optimized Data Center)**
A POD is a collection of computer systems, switch fabrics, storage systems, etc. with its own cooling, power feed and management interface. A POD resides in a single data center. A data center can include multiple PODs. A POD realizes an operational environment to host software applications and services. 

- **Data Fabric**
The Data Fabric is the physical infrastructure of one POD (Performance Optimized Data Center) carrying all traffic that is originated or terminated in a compute node except for out-of-band control traffic. The Data Fabric can also provide VIMs and vPOD infrastructure on these compute nodes with an in-band control network service. The Data Fabric HW comprises the HW switches and the cabling. 

- **Control Fabric**
The Control Fabric is the physical infrastructure realizing the Control Network within a datacenter which is used for out-of-band control and equipment management. 

- **Data Fabric Manager/Controller**
A manager/controller managing Data Fabric through its corresponding API/CLI.

**Reply by taylor** [reply]
See NSM and K8s glossaries below

**Reply by jeffsaelens** [reply]
Some of the terms we might want to revise:
Primary Network: This is describing the K8s overlay network (assuming you aren't peering with the underlay), so we should come up with a more specific K8s-centric name.
Secondary Network: Same issue as primary.

A lot of network word usage might be better suited with plane or overlay/underlay verbiage.

POD here happens to have overlap with a K8s term.

We should distinguish between control and management when talking about planes.

**Reply by mardim91** [reply]
@jeffsaelens hello,

do you think that the below definitions fit well to the different planes that we have in the infrastructure ?

**Management Plane**

Is the system entity which provides configuration, monitoring and management related capabilities for all the different levels of the infrastructure. 

**Control Plane**

Is the element of the system which provides control and signaling information necessary for the undisruptive flow of  data among the different infrastructure nodes. 

**Data Plane**

Is the part of the system where all the forwarding decisions are made for the steering of packets to their final destination.

**Reply by electrocucaracha** [reply]
Does _primary_ mean _mandatory_? In multus, it's possible to use any secondary network as the unique network for that pod so in that sense _primary_ is a concept relative to the pod.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example
  annotations:
    v1.multus-cni.io/default-network: secondary_net
```

**Comment by taylor** [comment]
NSM glossary https://docs.google.com/document/d/1zQOIQySDPgk2uUUKAWW34GiO0GVOwsN0qQA9tvSn97A/edit

**Comment by taylor** [comment]
K8s Plumbing Group definitions + terminology https://github.com/k8snetworkplumbingwg/multi-net-spec

v1.2

2.3 Network Attachment
A means of allowing a pod to directly communicate with a given logical or physical network.
Typically (but not necessarily) each attachment takes the form of a kernel network interface
placed in to the pod’s network namespace. Each attachment may result in zero or more IP
addresses being assigned to the pod.

**Comment by taylor** [comment]
**Kube-native**
Designed to run on Kubernetes...
TBD

**Comment by gupta-alok** [comment]
Based on previous comments, trying to make an attempt of improvement by revisiting following definitions:

- **Network Attachment**
A means of allowing a pod to directly communicate with a given logical or physical network. Typically (but not necessarily) each attachment takes the form of a kernel network interface placed in to the pod’s network namespace.  Each attachment may result in zero or more IP addresses being assigned to the pod.
**Ref**: https://github.com/k8snetworkplumbingwg/multi-net-spec/blob/a8b8b0c598a894457fa67e4467a2c47d46852b3d/v1.2/%5Bv1.2%5D%20Kubernetes%20Network%20Custom%20Resource%20Definition%20De-facto%20Standard.pdf
  - **Primary Network Attachment**
Each pod has a mandatory primary network attachment. This kernel network interface (eth0) is injected by the default CNI.
  - **Secondary Network Attachment**
On top of the primary network attachment a K8s pod may have zero or more secondary network attachments. These are requested in the pod specification by means of network annotations referring to Network Attachment Definitions (NADs).
A meta CNI plugin, such as Multus, iterates over the pod’s network annotations and invokes the secondary CNI plugin specified in each NAD to inject the requested secondary network attachment into the pod.
Secondary network attachments can take many forms, among others kernel network interfaces, PCI devices configured for use with DPDK, AF_XDP sockets, or user-space interfaces like memif.
- **Kubernetes Primary Network**
The network to which the K8s pods attach with their primary network interface (eth0). It includes the internal network interconnecting the K8s pods and any external networks with connectivity to the primary pod interfaces, either directly through IP routing, involving NAT and/or load-balancing. Today, K8s services reside on and are limited to the primary network.
- **Kubernetes Secondary Network**
A network, isolated from the primary network, that K8s pods can attach to through secondary network attachments. Typically, secondary networks have external connectivity through the DC Edge.
- **Management Plane**
Is the system entity which provides configuration, monitoring and management related capabilities for all the different levels of the infrastructure. 
- **Control Plane**
Is the element of the system which provides control and signaling information necessary for the undisruptive flow of data among the different infrastructure nodes.
- **Data Plane**
Is the part of the system where all the forwarding decisions are made for the steering of packets to their final destination.
- **Overlay Network**
Overlay Network is a networking abstraction that creates multiple layers of networking on top of physical network. Through overlay networks, pods can communicate with each other within their cluster and with the gateway(s) connected to networks outside of the data center by creating virtual networks.
- **External Network**
A network connecting the K8s cluster nodes and pods with external entities, typically the DC Edge for connectivity with the WAN.
- **Tenant Network**
The networks that carry tenant traffic (e.g. traffic to/from CNFs) are simply called Tenant Traffic Network. Tenant networks comprise primary and secondary networks. Tenant networks are typically isolated from infrastructure networks, such as the control network carrying the K8s cluster control plane or storage traffic.
- **DC-Gw/Site Router/DC Edge Router (PE + GW)**
The Data Center Gateway is a router through which the traffic entering or leaving to the data center, is routed. It faces on one side the data center network and on the other side a transport network. It may act as the DC Edge and may also be combined with DC Management Edge, WAN Edge and/or access edge. 
- **Data Fabric**
The Data Fabric is the physical network infrastructure of one datacenter carrying all traffic that is originated or terminated in a compute node except for out-of-band control traffic. The Data Fabric can also provide VIMs and virtual datacenter infrastructure on these compute nodes with an in-band control network service. The Data Fabric HW comprises the HW switches and the cabling. 
- **Control Fabric**
The Control Fabric is the physical infrastructure realizing the Control Network within a datacenter which is used for out-of-band control and equipment management. 
- **Data Fabric Manager/Controller**
A manager/controller managing Data Fabric through its corresponding API/CLI.

Cc: @jeffsaelens @taylor @tliron
